File: kafka-streams-scala/9-processor-api-scala/src/main/scala/com/github/simplesteph/udemy/kafka/streams/internal/Generator.scala
package com.github.simplesteph.udemy.kafka.streams.internal

import java.time.Instant

import com.github.simplesteph.{Character, Coupon, Game, Player, Reward, Victory}
import org.scalacheck.Gen

import scala.util.Random


trait Generator {

  private val sfCharacters: Array[Character] = Array(
    Character("Ken", "US"),
    Character("Ryu", "Japan"),
    Character("Geki", "Japan"),
    Character("Chun-Li", "China"),
    Character("Akuma", "Japan"),
    Character("Sakura", "Japan"),
    Character("Dhalsim", "India"),
    Character("Blair", "UK")
  )

  private val takkenCharacters: Array[Character] = Array(
    Character("Jin", "Japan"),
    Character("Asuka", "Japan"),
    Character("Emilie", "Monaco"),
    Character("Kazuya", "Japan")
  )

  private val kofCharacters: Array[Character] = Array(
    Character("Mai", "Japan"),
    Character("Ramon", "Mexico"),
    Character("Nelson", "Brazil"),
    Character("Vanessa", "France")
  )

  private val scCharacters: Array[Character] = Array(
    Character("Kilik", "China"),
    Character("Ivy", "UK"),
    Character("Siegfried", "HRE"),
    Character("Nightmare", "X")
  )

  private val smCharacters: Array[Character] = Array(
    Character("Galford", "US"),
    Character("Charlotte", "France"),
    Character("Haohmaru", "Japan"),
    Character("Ukyo Tachibana", "Japan")
  )

  private val CharactersMap = Map(
    Game.KingOfFighters -> kofCharacters,
    Game.SamuraiShodown -> smCharacters,
    Game.StreetFighter -> sfCharacters,
    Game.SoulCalibur -> scCharacters,
    Game.Takken -> takkenCharacters
  )

  val victoryGen: Gen[Victory] = for {
    terminalId <- Gen.choose(0, 9)

    game <- Gen.oneOf(Game.values())

    character <- Gen.oneOf(CharactersMap(game))

    player <- Gen.frequency((4, Player.PlayerOne), (1, Player.PlayerTwo))

  } yield Victory(game, terminalId.toString, character, Instant.now(), player)

  protected def nextVictory: Option[Victory] = victoryGen.sample

  protected def nextRewards(bound: Int): Vector[Reward] = (0 to bound).map { termId =>
    Reward(
      termId.toString,
      coupons = (0 until 5) // <- 5 coupons for each machine
        .map { couponId =>
        Coupon(s"Coupon#$termId.0.$couponId", availability = true)
      }
    )
  }.toVector

  protected def nextValidStatus: Boolean = Random.nextBoolean()
}

File: kafka-streams-java/8-suppress-operator/src/main/java/com/github/simplesteph/udemy/kafka/streams/internal/Generator.java
package com.github.simplesteph.udemy.kafka.streams.internal;

import com.github.simplesteph.Character;
import com.github.simplesteph.Game;
import com.github.simplesteph.Victory;
import org.jeasy.random.EasyRandom;
import org.jeasy.random.EasyRandomParameters;

import java.time.Instant;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static org.jeasy.random.util.CollectionUtils.randomElementOf;

public abstract class Generator {

    private static final String DAMAGED_TERMINAL = "7";

    private static final ArrayList<Character> sfCharacters = new ArrayList<>(Arrays.asList(
            new Character("Ken", "US"),
            new Character("Ryu", "Japan"),
            new Character("Geki", "Japan"),
            new Character("Chun-Li", "China"),
            new Character("Akuma", "Japan"),
            new Character("Sakura", "Japan"),
            new Character("Dhalsim", "India"),
            new Character("Blair", "UK")
    ));

    private static final ArrayList<Character> takkenCharacters = new ArrayList<>(Arrays.asList(
            new Character("Jin", "Japan"),
            new Character("Asuka", "Japan"),
            new Character("Emilie", "Monaco"),
            new Character("Kazuya", "Japan")
    ));

    private static final ArrayList<Character> kofCharacters = new ArrayList<>(Arrays.asList(
            new Character("Mai", "Japan"),
            new Character("Ramon", "Mexico"),
            new Character("Nelson", "Brazil"),
            new Character("Vanessa", "France")
    ));

    private static final ArrayList<Character> scCharacters = new ArrayList<>(Arrays.asList(
            new Character("Kilik", "China"),
            new Character("Ivy", "UK"),
            new Character("Siegfried", "HRE"),
            new Character("Nightmare", "X")
    ));

    private static final ArrayList<Character> smCharacters = new ArrayList<>(Arrays.asList(
            new Character("Galford", "US"),
            new Character("Charlotte", "France"),
            new Character("Haohmaru", "Japan"),
            new Character("Ukyo Tachibana", "Japan")
    ));

    private static final HashMap<Game, List<Character>> CHARACTERS_MAP = new HashMap<Game, List<Character>>() {{
        put(Game.KingOfFighters, kofCharacters);
        put(Game.SamuraiShodown, smCharacters);
        put(Game.StreetFighter, sfCharacters);
        put(Game.SoulCalibur, scCharacters);
        put(Game.Takken, takkenCharacters);
        put(Game.None, Collections.singletonList(new Character("X", "")));
    }};

    private static EasyRandomParameters parameters = new EasyRandomParameters()
            .seed(60L)
            .objectPoolSize(30)
            .ignoreRandomizationErrors(true)
            .randomize(Victory.class, () -> {
                        Random random = new Random();
                        Game game = nextGame();
                        Character character = nextCharacter(game);

                        // is this a late event?
                        boolean lateStatus = game == Game.None && random.nextBoolean();

                        // if it's late, the delay may be between 5 and 15 seconds
                        Instant eventTime =
                                lateStatus ? Instant.now().minusSeconds(15 - random.nextInt(10)) : Instant.now();

                        return new Victory(
                                game,
                                DAMAGED_TERMINAL,
                                character,
                                eventTime,
                                lateStatus
                        );
                    }
            );

    private static EasyRandom easyRandom = new EasyRandom(parameters);

    private static Game nextGame() {
        return randomElementOf(
                Stream.concat(
                        Arrays.stream(Game.values()),
                        Collections.nCopies(3, Game.None).stream()
                ).collect(Collectors.toList())
        );
    }

    private static Character nextCharacter(Game game) {
        return randomElementOf(CHARACTERS_MAP.get(game));
    }

    protected static Optional<Victory> nextVictory() {
        return Optional.ofNullable(easyRandom.nextObject(Victory.class));
    }
}
